File: frontend/src/types/bill.ts
import type { RateType, PaymentType } from './common';
import type {
  CustomerBillPaymentResponse,
  CreateCustomerBillPaymentRequest,
} from './customer-bill-payment';

/**
 * Bill item interface for internal use
 */
export interface BillItem {
  id?: string;
  billId?: string;
  productId: string;
  productName?: string;
  quantity: number;
  rate: number;
  discount?: number;
  taxPercentage: number;
  totalAmount?: number;
  description?: string;
}

/**
 * Bill interface for internal use
 */
export interface Bill {
  id: string;
  pumpMasterId: string;
  billNo: number;
  billDate: string;
  customerId: string;
  customerName?: string;
  rateType: RateType;
  billItems: BillItem[];
  totalAmount?: number;
  discountAmount?: number;
  taxAmount?: number;
  netAmount?: number;
  createdAt?: string;
  updatedAt?: string;
}

/**
 * Response DTO for Bill information - matches backend BillResponse
 */
export interface BillResponse {
  /** Unique identifier */
  id: string;
  /** Pump Master ID */
  pumpMasterId: string;
  /** Bill number */
  billNo: number;
  /** Bill date */
  billDate: string;
  /** Customer ID */
  customerId: string;
  /** Customer name */
  customerName?: string;
  /** Rate type */
  rateType: RateType;
  /** Total amount */
  totalAmount: number;
  /** Discount amount */
  discountAmount: number;
  /** Tax amount */
  taxAmount: number;
  /** Net amount */
  netAmount: number;
  /** Creation timestamp */
  createdAt: string;
  /** Last update timestamp */
  updatedAt: string;
  /** List of bill items */
  billItems: BillItemResponse[];

  payments?: CustomerBillPaymentResponse[];
}

/**
 * Response DTO for Bill Item information
 */
export interface BillItemResponse {
  id: string;
  billId: string;
  productId: string;
  productName?: string;
  hsnCode?: string;
  salesUnit?: string;
  quantity: number;
  rate: number;
  amount: number;
  gst: number;
  discount: number;
  netAmount: number;
}

export interface CreateBillItemRequest {
  productId: string;
  quantity: number;
  rate: number;
  gst?: number;
  discount?: number;
}

export interface CreateBillRequest {
  pumpMasterId: string;
  billNo: number;
  billDate: string;
  customerId: string;
  paymentType: PaymentType;
  rateType: RateType;
  billItems: CreateBillItemRequest[];
  discountAmount?: number;
  payments?: CreateCustomerBillPaymentRequest[];
}

export interface UpdateBillRequest {
  billNo?: number;
  billDate?: string;
  customerId?: string;
  rateType?: RateType;
  billItems?: CreateBillItemRequest[];
}

File: backend/src/main/java/com/reallink/pump/entities/Bill.java
package com.reallink.pump.entities;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.HashSet;
import java.util.Set;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.ForeignKey;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.Digits;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "pump_bill_master", uniqueConstraints = {
    @UniqueConstraint(name = "uk_bill_no_pump", columnNames = {"bill_no", "pump_master_id"})
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Bill extends BaseEntity {

    @NotNull(message = "Pump master is required")
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "pump_master_id", nullable = false, foreignKey = @ForeignKey(name = "fk_bill_pump_master"))
    private PumpInfoMaster pumpMaster;

    @NotNull(message = "Bill number is required")
    @Min(value = 1, message = "Bill number must be positive")
    @Column(name = "bill_no", nullable = false)
    private Long billNo;

    @NotNull(message = "Bill date is required")
    @Column(name = "bill_date", nullable = false)
    private LocalDate billDate;

    @NotNull(message = "Customer is required")
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "customer_id", nullable = false, foreignKey = @ForeignKey(name = "fk_bill_customer"))
    private Customer customer;

    @NotNull(message = "Bill type is required")
    @Enumerated(EnumType.STRING)
    @Column(name = "bill_type", nullable = false)
    private BillType billType;

    @NotNull(message = "Rate type is required")
    @Enumerated(EnumType.STRING)
    @Column(name = "rate_type", nullable = false)
    private RateType rateType;

    @NotNull(message = "Payment type is required")
    @Enumerated(EnumType.STRING)
    @Column(name = "payment_type", nullable = false)
    private PaymentType paymentType;

    @NotNull(message = "Total amount is required")
    @DecimalMin(value = "0.0", inclusive = true, message = "Total amount must be non-negative")
    @Digits(integer = 10, fraction = 2, message = "Total amount must have at most 10 digits and 2 decimal places")
    @Column(name = "total_amount", nullable = false, precision = 12, scale = 2)
    private BigDecimal totalAmount = BigDecimal.ZERO;

    @DecimalMin(value = "0.0", inclusive = true, message = "Discount amount must be non-negative")
    @Digits(integer = 10, fraction = 2, message = "Discount amount must have at most 10 digits and 2 decimal places")
    @Column(name = "discount_amount", precision = 12, scale = 2)
    private BigDecimal discountAmount = BigDecimal.ZERO;

    @DecimalMin(value = "0.0", inclusive = true, message = "Tax amount must be non-negative")
    @Digits(integer = 10, fraction = 2, message = "Tax amount must have at most 10 digits and 2 decimal places")
    @Column(name = "tax_amount", precision = 12, scale = 2)
    private BigDecimal taxAmount = BigDecimal.ZERO;

    @NotNull(message = "Net amount is required")
    @DecimalMin(value = "0.0", inclusive = true, message = "Net amount must be non-negative")
    @Digits(integer = 10, fraction = 2, message = "Net amount must have at most 10 digits and 2 decimal places")
    @Column(name = "net_amount", nullable = false, precision = 12, scale = 2)
    private BigDecimal netAmount = BigDecimal.ZERO;

    @Column(name = "vehicle_no")
    private String vehicleNo;

    @Column(name = "driver_name")
    private String driverName;

    @OneToMany(mappedBy = "bill", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<BillItem> billItems = new HashSet<>();

    @OneToMany(mappedBy = "bill", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<CustomerBillPayment> customerBillPayments = new HashSet<>();
}
